consensus/src/liveness/leader_reputation.rs

```diff
@@ -92,26 +110,32 @@ impl AptosDBBackend {
         &self,
         target_epoch: u64,
         target_round: Round,
-        events: &Vec<NewBlockEvent>,
+        events: &Vec<VersionedNewBlockEvent>,
         hit_end: bool,
-    ) -> Vec<NewBlockEvent> {
-        let has_larger = events.first().map_or(false, |e| {
-            (e.epoch(), e.round()) >= (target_epoch, target_round)
-        });
-        if !has_larger {
-            // error, and not a fatal, in an unlikely scenario that we have many failed consecutive rounds,
-            // and nobody has any newer successful blocks.
-            error!(
-                "Local history is too old, asking for {} epoch and {} round, and latest from db is {} epoch and {} round! Elected proposers are unlikely to match!!",
-                target_epoch, target_round, events.first().map_or(0, |e| e.epoch()), events.first().map_or(0, |e| e.round()))
+    ) -> (Vec<NewBlockEvent>, HashValue) {
+        // Do not warn when round==0, because check will always be unsure of whether we have
+        // all events from the previous epoch. If there is an actual issue, next round will log it.
+        if target_round != 0 {
+            let has_larger = events.first().map_or(false, |e| {
+                (e.event.epoch(), e.event.round()) >= (target_epoch, target_round)
+            });
+            if !has_larger {
+                // error, and not a fatal, in an unlikely scenario that we have many failed consecutive rounds,
+                // and nobody has any newer successful blocks.
+                error!(
+                    "Local history is too old, asking for {} epoch and {} round, and latest from db is {} epoch and {} round! Elected proposers are unlikely to match!!",
+                    target_epoch, target_round, events.first().map_or(0, |e| e.event.epoch()), events.first().map_or(0, |e| e.event.round()))
+            }
         }
 
+        let mut max_version = 0;
         let mut result = vec![];
         for event in events {
-            if (event.epoch(), event.round()) <= (target_epoch, target_round)
+            if (event.event.epoch(), event.event.round()) <= (target_epoch, target_round)
                 && result.len() < self.window_size
             {
-                result.push(event.clone());
+                max_version = std::cmp::max(max_version, event.version);
```

Comment:
I don't think we need this, the max version should come from the first/last event unless something is wrong